File: Gwanghun/공통문제/10주차/프로그래머스/3/42861. 섬 연결하기/섬 연결하기.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int parent[101];

bool isGreater(vector<int>v1, vector<int>v2){
    return v1[2]<v2[2];
}

int find(int x){
    if(parent[x]==x) return x;
    return parent[x]=find(parent[x]);
}

void unite(int a, int b){
    a=find(a);
    b=find(b);
    
    if(a!=b){
        parent[b]=a;
    }
}

int kruskal(int n, vector<vector<int>> costs) {
    int answer=0;
    int cnt=0;
    
    for(int i=0; i<n; i++) parent[i]=i;
    
    sort(costs.begin(), costs.end(), isGreater);
    
    for(auto edge : costs){
        int a = edge[0];
        int b = edge[1];
        int cost = edge[2];
        
        if(find(a)!=find(b)){
            unite(a,b);
            answer+=cost;
            cnt++;
            
            if(cnt==n-1) break;
        }
    }
    
    return answer;
    
}

int solution(int n, vector<vector<int>> costs) {
    int answer = kruskal(n,costs);
    return answer;
}

File: Choonghun/공통문제/10주차/프로그래머스 42861 - 섬 연결하기/섬 연결하기.py
def solution(n, costs):
    # 크루스칼 알고리즘 (최소 스패닝 트리 -> 각 노드를 모두 잇는 최소한의 간선 비용 구하기 문제)
    # 유사 알고리즘은 프림 알고리즘 (우선순위 큐 활용)이 있다.
    
    costs = [[c[2],c[0],c[1]] for c in costs] # 정렬을 위해 비용을 맨 앞으로 이동
    parents = [i for i in range(n+1)] # 부모 노드를 자신으로 등록
    
    def find(node):
        if parents[node] == node: #만약 node의 부모가 자기 자신이면 자신을 반환
            return node
        parents[node] = find(parents[node]) # 최상위 노드 찾기
        return parents[node]
        
    def union_node(x,y): # 트리 합치기
        px = find(x) # x와 y의 부모를 찾아서
        py = find(y)
         
        if px < py: # px가 더 작으면 py의 부모가 px
            parents[py] = px
        else: # py가 더 작으면 px의 부모가 py이다.
            parents[px] = py
    
    costs.sort() # 간선 배열 정렬
    answer = 0
    for c, x, y in costs:
        if find(x) != find(y): # 둘의 부모가 같다는 것은 속한 트리가 같다는 것이다.
            union_node(x,y)
            answer += c
    # print(parents)
    
    return answer